vecmath/include/vecmath/mat.h

```diff
@@ -786,6 +794,109 @@ namespace vm {
         }
         return result;
     }
+
+    /**
+     * Finds an LUP decomposition of matrix a.
+     *
+     * Give A, finds P,L,U satisfying PA=LU where P is a permutation matrix,
+     * where L is lower-triangular with the diagonal elements set to 1,
+     * U is upper-triangular.
+     *
+     * The permutation matrix is returned in a compressed form where each element of the vector represents a row of
+     * the permutation matrix, and a value of `i` means the `i`th column of that row is set to 1.
+     *
+     * From "LUP-Decomposition", Introduction to Algorithms by Cormen et. al., 2nd. ed. p752.
+     *
+     * @tparam T the component type
+     * @tparam S the number of components
+     * @param a the matrix to decompose
+     * @return {true, L and U packed into a single matrix, compressed permutation matrix}
+     *         or {false, unspecified, unspecified} if a decomposition doesn't exist.
+     */
+    template <typename T, size_t S>
+    std::tuple<bool, mat<T,S,S>, vec<size_t,S>> lupDecomposition(mat<T,S,S> a) {
```

Comment:
I would rename this to `findLUPDecomposition`; I prefer having a verb in function names.

Reply:
Or rather, `lupFindDecomposition` to be consistent with the other functions.